feat: support pyclass on tuple enums (#4072)
PyO3/pyo3: guide/src/class.md

    // ...
}

> /// PyO3 also supports enums with Struct and Tuple variants

Reviewer: Why change these to a doc comment, it's not like this would be a proper doc comment for the `enum Shape`.

> /// These complex enums have sligtly different behavior from the simple enums above

Reviewer: While at it:

```suggestion
/// These complex enums have slightly different behavior from the simple enums above
```

/// They are meant to work with instance checks and match statement patterns
/// The variants can be mixed and matched
/// Struct variants have named fields while tuple enums generate generic names for fields in order _0, _1, _2, ...
/// Apart from this both types are functionally identical
#[pyclass]
enum Shape {
    Circle { radius: f64 },

>     Rectangle{ height : f64, width : f64},

Reviewer: ```suggestion
    Rectangle { height: f64, width: f64 },
```

>     Square(u32),

Reviewer: Should be an f64 here too.

    Nothing(),
}
```

> 

Reviewer: Remove empty lines

> enum Shape {

Reviewer: This enum was the same in both instances, while now the first example has Square instead of RegularPolygon. I think it's best to keep it the same even with the tuple variant example.

    Circle { radius: f64 },
    Rectangle { width: f64, height: f64 },

>     RegularPolygon ( u32, f64 ),

Reviewer: ```suggestion
    RegularPolygon(u32, f64),
```

    Nothing { },
}

# #[cfg(Py_3_10)]
Python::with_gil(|py| {
    let circle = Shape::Circle { radius: 10.0 }.into_py(py);

>     let square = Shape::RegularPolygon ( 4, 10.0 ).into_py(py);

Reviewer: ```suggestion
    let square = Shape::RegularPolygon(4, 10.0).into_py(py);
```

        assert isinstance(square, cls)
        assert isinstance(square, cls.RegularPolygon)

>         assert square._0 == 4
>         assert square._1 == 10.0

Reviewer 2: Thank you very much for working on this! Just some initial thought, without having looked into the code yet.

I think I would find it more intuitive if tuple variants would use indexing on the Python side. 
```suggestion
        assert square[0] == 4
        assert square[1] == 10.0
```
This would be similar to how a normal tuple in Python would be accessed:
```python
square = (4, 10)
assert square[0] == 4
assert square[1] == 10
```
It might also solve the identifier problem you're having. What do you think about that? (`_0` does also use the "private" field convention, which is maybe not ideal, given that it is a normal public field)

Reviewer 2: On second thought, we could also have that _additionally_ such that matching still works.

Reviewer 2: I think generally we should advertise the tuple variants using indexing (here) and tuple match (below). Personally I think it looks nicer and in general I believe it is more intuitive and less confusing than referring to (kind of) arbitrary generated names.

We could definitely add a section (later) that mention these alternatives as well, but for the first introduction I think the above if more appropriate.

Reviewer 2: Perfect 💯

An enum is complex if it has any non-unit (struct or tuple) variants.

> Currently PyO3 supports only struct and tuple variants in a complex enum. Support for unit variants is planned.

Reviewer 3: As per https://github.com/PyO3/pyo3/issues/3749 I don't think we intend to make unit variants work any time soon.

```suggestion
PyO3 supports only struct and tuple variants in a complex enum. Unit variants aren't supported at present (the recommendation is to use an empty tuple enum instead).
```

        assert isinstance(square, cls)
        assert isinstance(square, cls.RegularPolygon)
        assert square._0 == 4
        assert square._1 == 10.0

        def count_vertices(cls, shape):
            match shape:
                case cls.Circle():
                    return 0
                case cls.Rectangle():
                    return 4

>                 case cls.RegularPolygon(_0=n):

Reviewer 2: I think we just need to generate the equivalent of
```rust
#[classattr]
const __match_args__: (&'static str, &'static str) = ("_0", "_1");
```

At least from a quick test with:
```rust
#[pyclass]
pub struct PyClass(i8, String);

#[pymethods]
impl PyClass {
    #[classattr]
    #[allow(non_upper_case_globals)]
    const __match_args__: (&'static str, &'static str) = ("_0", "_1");

    #[new]
    fn new(_0: i8, _1: String) -> Self {
        Self(_0, _1)
    }
}
```
Shows this Python code work fine:
```python
foo = PyClass(0, "Hello")
match foo:
    case PyClass(a, b):
        print(a, b)
```
(And without `__match_args__` it errors)

`impl_py_class_attribute` might come in handy to generate this as part of the macro
